### PR TITLE
Feat/MET-5383 Add rate control to validation workflow

### PR DESCRIPTION
**Pull request by JoanaCMS**

### CODE REVIEW

**Review by jeortizquan — COMMENTED** [review]
I works as intended, tested on local.
Just have small observations.

File: pom.xml [hunk 1]
```diff
@@ -62,8 +62,28 @@
...
+    <dependency>
+      <groupId>com.bucket4j</groupId>
+      <artifactId>bucket4j-core</artifactId>
+      <version>${version.bucket4j}</version>
+    </dependency>
+    <dependency>
+      <groupId>com.bucket4j</groupId>
+      <artifactId>bucket4j-postgresql</artifactId>
+      <version>${version.postgresql.bucket4j}</version>
+      <exclusions>
```
[the remark this answers is not in the document]
Comment (JoanaCMS) [reply pom.xml:79]: It conflicted with `HikariCP` that came with the `org.springframework.boot:spring-boot-starter-jdbc:jar:2.7.11` dependency

File: src/main/java/eu/europeana/metis/sandbox/controller/ratelimit/RateLimitInterceptor.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
+package eu.europeana.metis.sandbox.controller.ratelimit;
+
+
+import io.github.bucket4j.*;
```
[the remark this answers is not in the document]
Comment (JoanaCMS) [reply src/main/java/eu/europeana/metis/sandbox/controller/ratelimit/RateLimitInterceptor.java:4]: Fixed

File: src/main/java/eu/europeana/metis/sandbox/controller/ratelimit/RateLimitInterceptor.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        this.capacity = capacity;
+        this.postgreSQLManager = postgreSQLManager;
+        bucketConfiguration = BucketConfiguration.builder()
+                .addLimit(Bandwidth.classic(capacity, Refill.intervally(capacity, Duration.ofSeconds(time))))
+                .build();
+    }
+
+    @Override
+    public boolean preHandle(HttpServletRequest request, @NotNull HttpServletResponse response, @NotNull Object handler) throws Exception {
+        Long key = (long) request.getRemoteAddr().hashCode();
```
[the remark this answers is not in the document]
Comment (JoanaCMS) [reply src/main/java/eu/europeana/metis/sandbox/controller/ratelimit/RateLimitInterceptor.java:42]: Done

File: src/main/java/eu/europeana/metis/sandbox/controller/ratelimit/RateLimitInterceptor.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        ConsumptionProbe probe = resolveBucket(key);
```
[the remark this answers is not in the document]
Comment (JoanaCMS) [reply src/main/java/eu/europeana/metis/sandbox/controller/ratelimit/RateLimitInterceptor.java:43]: Done

File: src/main/java/eu/europeana/metis/sandbox/controller/ratelimit/RateLimitInterceptor.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+        response.addHeader("X-Rate-Limit-Limit", String.valueOf(capacity));
+        if (probe.isConsumed()) {
+            response.addHeader("X-Rate-Limit-Remaining", String.valueOf(probe.getRemainingTokens()));
+            return true;
+        } else {
+            long waitForRefill = probe.getNanosToWaitForRefill() / 1_000_000_000;
```
[the remark this answers is not in the document]
Comment (JoanaCMS) [reply src/main/java/eu/europeana/metis/sandbox/controller/ratelimit/RateLimitInterceptor.java:49]: Done

File: src/main/java/eu/europeana/metis/sandbox/controller/ratelimit/RateLimitInterceptor.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+            response.addHeader("X-Rate-Limit-Reset", String.valueOf(waitForRefill));
+            response.sendError(HttpStatus.TOO_MANY_REQUESTS.value(),
+                    "You have exhausted your API Request Quota");
+            return false;
+
+        }
+    }
+
+    private ConsumptionProbe resolveBucket(Long apiKey) {
+        Request<ConsumptionProbe> request = new Request<>(new TryConsumeAndReturnRemainingTokensCommand(1), null, null);
```
[the remark this answers is not in the document]
Comment (JoanaCMS) [reply src/main/java/eu/europeana/metis/sandbox/controller/ratelimit/RateLimitInterceptor.java:59]: Done

**Review by jeortizquan — APPROVED** [review]

**Pull request merged**